### ISSUE TITLE
CatBoostClassifier models are not the same between CPUs

### ISSUE THREAD

**Issue by nguyenbh1507** [catboost/catboost#1825]
Problem: CatBoostClassifier models are not the same between CPUs, while input data and parameters are the same.
catboost version: 0.26.1
Operating System: Ubuntu 18.04
CPU_1: Intel i7-7700
CPU_2: Intel Xeon
Input data: CSV, 28 float fields and 2 categorical fields
Label: binary 0, 1
parameters of CatBoostClassifier: all are by default except random_state=42

Hi there,

I am facing such a following problem:
I use CatBoostClassifier to develop an binary classifier and train it on two computers with Intel i7-7700 and Intel Xeon.
The parameters are all by default, except random_state=42 to get the reproducibility.
The training data are the same, saved in text-CSV format with 28 float fields and 2 categorical fields.
Data were loaded by Pandas 0.25.3

When I used .predict_proba() to get the probabilities on training data by two computers, 
I got two totally different results, not at 10^-16-10^-14 in some cases.
I tried to save model from 1st computer (Intel i7) and use it to predict on 2nd computer (Intel Xeon),
this gave me the same results.
In this problem, I think I cannot reproduce the same model between i7 and Xeon.
Does anyone face with this such problem?

I know that CatBoost yields different results between GPU/CPU and already experimented with it.

Could you please give me a suggestion to fix it?
Thank you very much!

**Comment by LyzhinIvan** [comment]
Hi, @nguyenbh1507!

Which exactly Xeon do you have?

**Comment by nguyenbh1507** [comment]
> Hi, @nguyenbh1507!
> 
> Which exactly Xeon do you have?
Thank you,
I will check it and share you.

**Comment by nguyenbh1507** [comment]
> @nguyenbh1507 Most likely you should disable AVX2 branch here
> 
> https://github.com/catboost/catboost/blob/581282e4c172de7ac0a9ffb5dd54a88d7f2b6942/library/cpp/fast_exp/fast_exp.cpp#L51


Hi @Evgueni-Petrov-aka-espetrov 
Thank you for your comment
I use Python package only.
How do I disable AVX2 branch in Python 3+?

**Comment by Evgueni-Petrov-aka-espetrov** [comment]
@nguyenbh1507 If this issue is a blocker for you, just fix source code, and rebuild the python package.
This is an emergency recipe, of course.

**Comment by nguyenbh1507** [comment]
Hello @Evgueni-Petrov-aka-espetrov 

Thank for your suggestion. I will try it with my collaborator, he's better than me in such problem.
By the way, I also tried another ways, the differences become less:
- increase depth  (max_depth) from default 6 to 8 (overfit on train data),
- decrease subsample ratio of columns when constructing each tree (colsample_bylevel/rms) from default 1 to 0.5

Still could not understand what happened.

**Comment by LyzhinIvan** [comment]
@nguyenbh1507, I close issue because @Evgueni-Petrov-aka-espetrov has answered your question. In addition to his words here is [the exact place in the source code](https://github.com/catboost/catboost/blob/master/library/cpp/fast_exp/fast_exp.cpp#L51) that should be modified to get the same results at all CPUs. Feel free to reopen if new questions will arise.
